README.md

MtreeRing
=======

<!--require(knitr);require(markdown);knit("README.Rmd")-->


**Authors:** [Jingning Shi](https://www.researchgate.net/profile/Jingning-Shi), [Wei Xiang](https://www.researchgate.net/profile/Wei-Xiang-11)<br/>
**License:** [GPL3](https://cran.r-project.org/web/licenses/GPL-3)

<!--pkg badges-->
[![TravisCI Build Status](https://api.travis-ci.org/ropensci/MtreeRing.svg?branch=master)](https://api.travis-ci.org/ropensci/MtreeRing.svg?branch=master)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/cti7i110hecl8kpf?svg=true)](https://ci.appveyor.com/project/JingningShi/MtreeRing)
[![codecov](https://codecov.io/github/ropensci/MtreeRing/coverage.svg?branch=master)](https://codecov.io/github/ropensci/MtreeRing?branch=master)
[![ropensci](https://badges.ropensci.org/287_status.svg)](https://github.com/ropensci/software-review/issues/287)
[![cran checks](https://cranchecks.info/badges/worst/MtreeRing)](https://cranchecks.info/pkgs/MtreeRing)
[![Downloads](https://cranlogs.r-pkg.org/badges/MtreeRing)](https://CRAN.R-project.org/package=MtreeRing)

`MtreeRing` is a tool for automatically measuring tree-ring width using image processing techniques.

## Installation

Install the stable version from CRAN


```r
install.packages("MtreeRing")
```

or the development version from GitHub


```r
# install.packages("devtools")
devtools::install_github("ropensci/MtreeRing")
```

## Ring-width measurement

### 1. Read an image


```r
library(MtreeRing)
## Read and plot a tree ring image
img.name <- system.file("001.png", package = "MtreeRing")
t1 <- ring_read(img = img.name, dpi = 1200, plot = TRUE)
```

`ring_read` supports commonly used image formats, including png, tiff, jpg and bmp.

### 2. Detect ring borders 

After plotting the image, the automatic detection of ring borders can be performed using three alternative methods: (1) watershed algorithm; (2) Canny edge detector; (3) a linear detection algorithm from R package [measuRing](https://CRAN.R-project.org/package=measuRing).


```r
## Split a long core sample into 2 pieces to
## get better display performance and use the
## watershed algorithm to detect ring borders:
t2 <- ring_detect(ring.data = t1, seg = 2, method = 'watershed')
```

<center><img src="inst/README-img001.png" width = "65%" height = "65%" /></center>
<center>Figure 1. The automatic detection of ring borders</center>

### 3. Calculate ring-width series 

If all ring borders are correctly identified, you can generate a ring-width series in data frame format. Use `write.rwl` to export the ring-width series to an rwl file.


```r
rw.df <- ring_calculate(ring.data = t2, seriesID = "940220")
library(dplR) # A dendrochronological analysis package
fn <- tempfile(fileext=".rwl")
write.rwl(rwl.df = rw.df, fname = fn, format = "tucson")
```


## Shiny application

If you are not familiar with R and its command line interface, the shiny-based app is a good alternative.


```r
MtreeRing::ring_app_launch()
```

This command allows to run a Shiny-based application within the system's default web browser. The app provides a beginner-friendly graphical interface and supports more flexible mouse-based interactions, allowing image files to be uploaded up to 150 MB in size.

The dashboard has three components: a header, sidebar and body, like this

<img src="inst/README-img002.png" width = "85%" height = "85%" />

A workflow for the Shiny app can be found at https://ropensci.github.io/MtreeRing/articles/app-MtreeRing.html. Most steps are demonstrated with a gif to make the workflow more understandable.

## Ring width correction

If an increment borer is used to extract samples, it is well known that the auger sometimes fails to traverse the pith of the sampled tree but passes through one side of the pith at a certain distance. Tangent lines of rings close to the pith are therefore not perpendicular to the horizontal path, which may lead to considerable errors in ring widths.

Under such conditions, you can create two paths by setting the argument `incline = TRUE`, or by ticking the checkbox **Inclined tree rings**. See this example.

<img src="inst/RingCorrection.png" width = "80%" height = "80%" /> 

The line segment connecting two dots on the same ring should match the tangent of a tree ring border. The corrected ring width is estimated from the distance between adjacent rings and orientation of ring borders.

## Support

Any feedback, bug reports or suggestions are welcomed. If you have a comment on `MtreeRing`, or you find a bug in the released or beta versions, please submit bugs and/or feature requests at https://github.com/ropensci/MtreeRing/issues. Include the package version, OS, and any command-line required to reproduce the problem.

## Code of conduct

I will try to add new features based on user feedback. It is hoped that others will contribute additional useful features. Please note that the 'MtreeRing' project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to this project, you agree to abide by its terms.

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
MtreeRing News
=======
### MtreeRing 1.4.5 (2021-04-20)
- Update CSS styles of the graphical user interface

### MtreeRing 1.4.4 (2021-03-15)
- Update links in the README file

### MtreeRing 1.4.3 (2021-03-14)
- Replace ‘spatstat’ with ‘spatstat.geom’ based on feedback from the Spatstat Team

### MtreeRing 1.4.2 (2019-10-03)
- Update GitHub links in `ring_app_launch` function

### MtreeRing 1.4.1 (2019-09-23)
- Update README file

### MtreeRing 1.4.0 (2019-09-01)
- Redesign the interface and measurement process for the Shiny app
- Update documentation for the redesigned Shiny app
- Add a Support section in the README file
- Replace all vignettes with GitHub links

### MtreeRing 1.3.1 (2019-05-29)
- Update README file

### MtreeRing 1.3 (2019-05-20)
- Add a NEWS.md file to track changes
- Add CODE_OF_CONDUCT.md
- Add `plot` parameter to `ring_read`
- Rename all exported functions using snake_case style
- Update documentation for the Shiny app and mouse actions
- Update README file
- Fix a bug where `ring_detect` fails to detect ring borders
- Remove intro vignette and add three new vignettes to demonstrate major functionality

### MtreeRing 1.2 (2019-03-02)
- Add appveyor-CI and travis-CI
- Add README file
- Add tests for all functions and Shiny app
- Add intro vignette
- Update documentation for `autoDetect`
- Update DESCRIPTION URL and bug report

### MtreeRing 1.1 (2018-10-16)
- Initial CRAN release


R/createUI.R
#' @importFrom shiny actionButton checkboxInput div fileInput conditionalPanel
#' downloadButton downloadHandler helpText icon numericInput observeEvent
#' plotOutput reactiveValues renderPlot renderTable selectInput shinyApp
#' sliderInput tableOutput tabPanel textInput updateActionButton brushOpts
#' dblclickOpts column
#' updateCheckboxInput updateTextInput hr br fluidRow
#' @importFrom shinydashboard dashboardHeader dashboardSidebar dashboardBody
#' dashboardPage sidebarMenu tabBox tabItem tabItems menuItem box
#' @importFrom  shinyWidgets prettySwitch prettyCheckbox prettyRadioButtons
#' radioGroupButtons sendSweetAlert updatePrettySwitch updatePrettyCheckbox
#' updatePrettyRadioButtons useSweetAlert colorSelectorInput

createUI <- function()
{
shiny.title <- dashboardHeader(title = 'MtreeRing')
shiny.sider <- dashboardSidebar(
  sidebarMenu(
    menuItem('Image Loading',tabName = 'input_pre', 
             icon = icon('folder-open', lib = 'font-awesome')),
    menuItem('Measurement',tabName = 'mea_arg', 
             icon = icon('gear', lib = 'font-awesome'))
  )
)
page1 <- fluidRow(
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%; 
                font-weight: bolder', 'Image Preview'),
    width = 12, status = 'primary', solidHeader = T, collapsible = T,
    plotOutput('pre.img',
               brush = brushOpts(
                 id = "plot1_brush",
                 opacity = 0.25,
                 resetOnNew = TRUE)
    )
  ), 
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%;
                font-weight: bolder', 'Image Upload'),
    width = 4, status = 'primary', solidHeader = T, collapsible = T,
    conditionalPanel(
      condition = '!input.inmethod',
      fileInput('select.file', 'Choose an image file',
                buttonLabel = 'Browse...', width = '80%')
    ),
    prettySwitch(inputId = "magick.switch", label = "Magick ON",
                 value = TRUE, fill = TRUE, status = "success"),
    helpText('Image upload is limited to 150 MB per file. Supported',
             ' formats include png, jpg, tif and bmp.',
             style = 'color:#000000;font-size:90%'),
    prettyCheckbox(
      inputId = "inmethod", 
      label = div(style = 'color:#000000;font-weight: bolder;', 'Image Path'), 
      shape = "curve", value = F, status = "success"),
    conditionalPanel(
      condition = 'input.inmethod',
      textInput('enter.path', 'Enter file path', ''),
      helpText('For example: C:/Users/shiny/img01.png',
               style = 'color:#000000;font-size:90%'),
      hr()
    ),
    actionButton(
      'buttoninputimage', 'Load ',
      class = "btn btn-primary btn-md",
      icon = icon('upload',  "fa-1x"),
      style = 'color:#FFFFFF;text-align:center;
              font-weight: bolder;font-size:110%;'),
    useSweetAlert()
  ),
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%;
                font-weight: bolder', 'Image Rotation'),
    width = 3, status = 'primary', solidHeader = T, collapsible = T,
    prettyRadioButtons(inputId = "rotatede", label = "",
                       choices = c("0 degrees" = "rotate0",
                                   "90 degrees" = "rotate90",
                                   "180 degrees" = "rotate180",
                                   "270 degrees" = "rotate270"),
                       shape = "curve", status = "success",
                       fill = TRUE, inline = TRUE),
    helpText("Rotation angle in degrees. Note that the bark ",
             "side should be placed at the left side of the ",
             "graphical window and the pith side at the right.",
             style = 'color:#000000;font-size:90%;text-align:justify;'),
    actionButton(
      'buttonrotate', 'Rotate',
      class = "btn btn-primary btn-md",
      icon = icon('repeat',"fa-1x"),
      style = 'color:#FFFFFF;text-align:center;
              font-weight: bolder;font-size:110%;')
  ),
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%;
                font-weight: bolder', 'Image Cropping'),
    width = 3, status = 'primary', solidHeader = T, collapsible = T,
    helpText("To remove unwanted cores and irrelevant objects, ",
             "move the mouse to the core you wish to measure and",
             "create a rectangle by brushing, see details below.",
             style = 'color:#000000;font-size:90%;text-align:justify;'),
    prettyRadioButtons(inputId = "cropcondition", label = "",
                       choiceNames = 'UNCROPPED', choiceValues = list('a'),
                       status = "danger", shape = "square",
                       fill = FALSE, inline = FALSE),
    prettyCheckbox(
      inputId = "showcropp", 
      label = div(style = 'color:#000000;font-weight: bolder;', 'Show Help'),
      shape = "curve", value = F, status = "success"
    ),
    conditionalPanel(
      condition = 'input.showcropp',
      helpText(
        "The operation \"brush\" allows users to create a transparent ", 
        "rectangle on the image and drag it around. For cores scanned ", 
        "side by side, the user can choose a core of interest by brushing.", 
        style = 'color:#000000;text-align:justify;'),
      helpText(
        "After brushing, click on the button \"Crop\" to create a",
        " cropped area. The measurement will be performed within", 
        " this area, rather than the whole (uncropped) image.",
        style = 'color:#000000;text-align:justify;'),
      helpText(
        "To cancel this operation, click on the button \"Cancel\".",
        " If the transparent rectangle exists, the user should first ",
        "click on the outer region of the rectangle (this will make the",
        " rectangle disappear) and then click on the button \"Cancel\".",
        style = 'color:#FF0000;text-align:justify;')
    ),  
    hr(),
    actionButton(
      'buttoncrop', 'Crop',
      class = "btn btn-primary btn-md",
      icon = icon('crop',"fa-1x"),
      style = 'color:#FFFFFF;text-align:center;
             font-weight: bolder;font-size:110%;')
  )
)
page2.1 <- fluidRow(
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%;
                font-weight: bolder', 'Options'),
    width = 4, status = 'primary', solidHeader = T, collapsible = T,
    textInput('tuid', 'Series ID', '', width = '75%'),
    textInput('dpi', 'DPI', '', '75%'),
    textInput('sample.yr', 'Sampling year', '', '75%'),
    textInput('m.line', 'Y-coordinate of path', '', '75%'),
    prettyCheckbox(
      inputId = "incline", 
      label = div(
        style = 'color:#000000;font-weight: bolder;', 'Inclined tree rings'), 
      shape = "curve", value = F, status = "success"
    ),
    conditionalPanel(
      condition = 'input.incline',
      numericInput('h.dis', 'Distance between paths (mm)', 
                   1, 0.1, 30, 0.1, width = '75%')
    ),
    br(),
    radioGroupButtons(
      inputId = "measuremethod", 
      label = 'Measurement mode',
      status = "btn btn-primary btn-md",
      #individual = T,
      size = 'normal',
      selected = 'auto',
      choiceNames = list(
          div(style = 'color:#FFFFFF;font-weight: bolder;', 'Manual'), 
          div(style = 'color:#FFFFFF;font-weight: bolder;', 'Automation')),
      choiceValues = list('manual', 'auto'),
      width = '100%') 
  ),
  box(
    title = div(style = 'color:#FFFFFF;font-size:80%;
                font-weight: bolder', 'Options'),
    width = 4, status = 'primary', solidHeader = T, collapsible = T,
    sliderInput('linelwd', 'Path width', 
                0.2, 3, 1, 0.1, width = '80%'),
    sliderInput('label.cex', 'Magnification for labels',
                0.2, 3, 1, 0.1, width = '80%'),
    radioGroupButtons(
      inputId = "pch", 
      label = 'Symbol for borders', status = "btn btn-primary btn-md",
      size = 'sm',
      choiceNames = list(
        div(style = 'color:#FFFFFF;font-weight: bolder;',
            icon('circle', 'fa-lg')), 
        div(style = 'color:#FFFFFF;font-weight: bolder;',
            icon('circle', 'fa-1x')), 
        div(style = 'color:#FFFFFF;font-weight: bolder;',
            icon('circle-o', 'fa-1x')), 
        div(style = 'color:#FFFFFF;font-weight: bolder;',
            icon('times', 'fa-1x')),
        div(style = 'color:#FFFFFF;font-weight: bolder;',
            icon('plus', 'fa-1x'))
      ),
      selected = '20', 
      choiceValues = list('19', '20', '1', '4', '3'),
      width = '100%'
    ),
    colorSelectorInput(
      inputId = "border.color", label = "Color for borders",
      choices = c("black", "gray", "white", "red", "#FF6000", 
                 "#FFBF00", "#DFFF00", "#80FF00", "#20FF00", 
                 "#00FF40", "#00FF9F", "cyan", "#009FFF", "#0040FF",
                 "#2000FF", "#8000FF", "#DF00FF", "#FF00BF"),
      selected = '#20FF00', mode = "radio", display_label = FALSE, ncol = 9
    ),
    colorSelectorInput(
      inputId = "label.color", label = "Color for labels",
      choices = c("black", "gray", "white", "red", "#FF6000", 
                 "#FFBF00", "#DFFF00", "#80FF00", "#20FF00", 
                 "#00FF40", "#00FF9F", "cyan", "#009FFF", "#0040FF",
                 "#2000FF", "#8000FF", "#DF00FF", "#FF00BF"),
      selected = 'black', mode = "radio", display_label = FALSE, ncol = 9
    )
  ),
  conditionalPanel(
    condition = 'input.measuremethod=="auto"',
    box(
      title = div(style = 'color:#FFFFFF;font-size:80%;
                  font-weight: bolder', 'Options'),
      width = 4, status = 'primary', solidHeader = T, collapsible = T,
      prettyCheckbox(
        inputId = "isrgb", 
        label = div(
          style = 'color:#000000;font-weight:bolder;', "Default RGB"), 
        shape = "curve", value = T, status = "success"
      ),
      conditionalPanel(
        condition = '!input.isrgb',
        textInput('customRGB', 'Custom RGB', '0.299,0.587,0.114'),
        helpText('Note:The three numbers correspond to',
                 'R, G and B components,respectively.',
                 style = 'color:#000000;font-weight: bolder'),
        hr()
      ),
      radioGroupButtons(
        inputId = "method",
        label = 'Ring detection method',
        status = "btn btn-primary btn-md",
        #individual = T,
        selected = 'canny',
        size = 'normal',
        choiceNames = list(
          div(style = 'color:#FFFFFF;font-weight: bolder;font-size:80%',
              'Watershed'),
          div(style = 'color:#FFFFFF;font-weight: bolder;font-size:80%',
              'Canny'),
          div(style = 'color:#FFFFFF;font-weight: bolder;font-size:80%',
              'measuRing')
        ),
        choiceValues = list('watershed', 'canny', 'lineardetect'), 
        width = '100%'
      ),
      conditionalPanel(
        condition = 'input.method=="watershed"',
        selectInput('watershed.threshold',
                    'Otsu threshold',
                    c('Auto (Recommended)' = 'auto',
                      'Custom' = 'custom.waterthr'),
                    width = '75%'
        ),
        conditionalPanel(
          condition = 'input["watershed.threshold"]=="auto"',
          sliderInput('watershed.adjust',
                      'Threshold adjusment factor',
                      0.5, 1.5, 0.8, 0.05, width = '85%')
        ),
        conditionalPanel(
          condition = 'input["watershed.threshold"]=="custom.waterthr"',
          textInput('watershed.threshold2', 'Threshold value', ''),
          'A value of the form XX% (e.g. 98%)'
        )
      ),
      conditionalPanel(
        condition = 'input.method=="canny"',
        prettyCheckbox(
          inputId = "defaultcanny", 
          label = div(
            style = 'color:#000000;font-weight: bolder;',
            "Auto threshold (Recommanded)"), 
          shape = "curve", value = T, status = "success"),
        conditionalPanel(
          condition = 'input.defaultcanny',
          sliderInput('canny.adjust',
                      'Threshold adjusment factor',
                      0.8, 1.8, 1.4, 0.05, width = '75%')
        ),
        conditionalPanel(
          condition = '!input.defaultcanny',
          textInput('canny.t2', 'Threshold for strong edges', '', '75%'),
          textInput('canny.t1', 'Threshold for weak edges', '', '75%')
        ),
        numericInput('canny.smoothing',
                     'Degree of smoothing',
                     1, 0, 4, 1, width = '75%')
      ),
      conditionalPanel(
        condition = 'input.method!="lineardetect"',
        prettyCheckbox(inputId = "defaultse", 
                       label = div(
                         style = 'color:#000000;font-weight: bolder;',
                         "Default structuring elements"), 
                       shape = "curve", value = T, status = "success"),
        conditionalPanel(
          condition = '!input.defaultse',
          textInput('struc.ele1', 'First structuring element', '', '75%'),
          textInput('struc.ele2', 'Second structuring element', '', '75%')
        ),
        hr()
      ),
      conditionalPanel(
        condition = 'input.method=="lineardetect"',
        textInput('origin', ' Origin in smoothed gray', '0', '75%'),
        'If you use the linear detection, don\'t ',
        'tick the checkbox "Inclined tree rings".',
        hr()
      ),
      helpText('Automatic detection may take a few seconds (depending',
               ' on the image size and complexity of the sample).',
               #style = 'color:#000000;font-size:95%;text-align:justify;')
               style = 'color:#000000;font-size:90%;')
    )
  ),
  box(
    title = div(style = 'color:#FFFFFF;font-size:100%;
                font-weight: bolder', 'Tree Ring Detection'),
    width = 12, status = 'primary', solidHeader = T, collapsible = T,
    conditionalPanel(
      condition = 'input.measuremethod!="auto"',
      actionButton(
        'buttoncreatpath', 'Create Path',
        class = "btn btn-primary btn-md", icon = icon('plus'),
        style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
      ),
      useSweetAlert(),
      actionButton(
        'buttonrcm', 'Remove All',
        class = "btn btn-danger btn-md", icon = icon('trash'),
        style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
      ),
      useSweetAlert()
    ),
    conditionalPanel(
      condition = 'input.measuremethod=="auto"',
      actionButton(
        'buttoncreatpath2', 'Create Path',
        class = "btn btn-primary btn-md", icon = icon('plus'),
        style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
      ),
      useSweetAlert(),
      actionButton(
        'buttonrcm2', 'Remove All',
        class = "btn btn-danger btn-md", icon = icon('trash'),
        style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
      ),
      useSweetAlert(),
      actionButton(
        'button_run_auto', 'Run Detection',
        class = "btn btn-success btn-md", icon = icon('play'),
        style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
      ),
      useSweetAlert()
    ),
    hr(),
    plotOutput('pre.img2',
               dblclick = "plot2_dblclick",
               brush = brushOpts(
                 id = "plot2_brush",
                 resetOnNew = TRUE
               )
    ),
    hr(),
    actionButton(
      'buttonsubimg', 'Create Sub-image',
      class = "btn btn-primary btn-md", icon = icon('search-plus'),
      style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
    )
  ),
  box(
    title = div(style = 'color:#FFFFFF;font-size:100%;
                font-weight: bolder', 'Tree Ring Editing'),
    width = 12, status = 'primary', solidHeader = T, collapsible = T,
    actionButton(
      'buttonzoomdel', 'Delete Border',
      class = "btn btn-danger btn-md",
      icon = icon('eraser'),
      style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
    ),
    useSweetAlert(),
    hr(),
    plotOutput('zoom.img',
               dblclick = dblclickOpts(id = "zoom_dblclick"),
               brush = brushOpts(
                 id = "zoom_brush",
                 resetOnNew = TRUE
               )
    )
  )
)
page2.2 <- fluidRow(
  column(width = 12,
    conditionalPanel(
      condition = '!input.tuheader',
      box(
        title = div(style = 'color:#FFFFFF;font-size:80%;
                    font-weight: bolder', 'Delete Borders'),
        width = 3, status = 'primary', solidHeader = T, collapsible = T,
        conditionalPanel(
          condition = 'input.incline',
          textInput('del.u', 'Border number in the upper portion', '', '75%'),
          textInput('del.l', 'Border number in the lower portion', '', '75%')
        ),
        conditionalPanel(
          condition = '!input.incline',
          textInput('del', 'Border number', '', '75%')
        ),
        helpText(
          "To perform a mass deletion of ring borders, use commas ",
          "to separate border numbers, e.g. 1, 2, 3, 4",
          style = 'color:#000000;text-align:justify;'
        ),
        br(),
        br(),
        actionButton(
          'button_del', 'Delete Border',
          class = "btn btn-danger btn-md", icon = icon('eraser'),
          style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
        )
      )
    ),
    tabBox(
      #title = tagList(shiny::icon("gear"), 'Output'),
      title = div(
        style = 'color:#000000;font-weight: bolder;',
        icon('cog', class = 'fa-spin', lib = 'font-awesome'), 'Output'),
      width = 6,
      tabPanel(
        div(style = 'color:#000000;font-weight: bolder;',
            icon('list-ol', 'fa-1x'), ' Results'),
        #HTML("<p style = 'color:#000000;'><b>Results</b></p>"),
        actionButton(
          'button_results', 'Generate Series',
          class = "btn btn-primary btn-md",
          style = 'color:#FFFFFF;text-align:center;font-weight:bolder;'
        ),
        useSweetAlert(),
        actionButton(
          'button_hide', 'Hide Series',
          class = "btn btn-primary btn-md",
          style = 'color:#FFFFFF;text-align:center;font-weight:bolder;'
        ),
        useSweetAlert(),
        br(),
        tableOutput('results')
      ),
      tabPanel(
        div(style = 'color:#000000;font-weight: bolder;',
            icon('arrow-down', 'fa-1x'), ' CSV'
        ),
        textInput('csv.name', 'Name of the csv file', '', width = '50%'),
        helpText(
          style = 'color:#000000;font-weight: normal;',
          'The filename extension is not required. ',
          'Leave blank to use the current series ID.'
        ),
        helpText(
          style = 'color:#FF0000;font-weight: normal;',
          'Attention: if running the app within an RStudio window',
          ', the rename operation doesn\'t work. Please run the app',
          ' within a browser.'
        ),
        hr(),
        #HTML("<p style = 'color:#000000;'><b>CSV</b></p>"),
        downloadButton(
          'RingWidth.csv', 'Download CSV',
          class = "btn btn-primary btn-md",
          style = 'color:#FFFFFF;text-align:center;font-weight:bolder;'
        )
      ),
      tabPanel(
        div(style = 'color:#000000;font-weight: bolder;',
            icon('arrow-down', 'fa-1x'), ' RWL'),
        textInput('rwl.name', 'Name of the rwl file', '', width = '50%'),
        helpText(style = 'color:#000000;font-weight: normal;',
                'The filename extension is not required. ',
                ' Leave blank to use the current series ID.'),
        helpText(style = 'color:#FF0000;font-weight: normal;',
                'Attention: if running the app within an RStudio window',
                ', the rename operation doesn\'t work. Please run the app',
                ' within a browser.'),
        hr(),
        selectInput('tuprec', 'Precision of the rwl file',
                    c('0.01' = '0.01', '0.001' = '0.001'),
                    selected = '0.01', width = '50%'),
        helpText(style = 'color:#000000;font-weight: normal;', 
                 'Units are in mm.'),
        hr(),
        checkboxInput('tuheader', 'Header of the File', F),
        conditionalPanel(  
           condition = 'input.tuheader',
           actionButton(
            'reset.hdr', 'Reset Header',
            class = "btn btn-danger btn-md",
            icon = icon('trash'),
            style = 'color:#FFFFFF;text-align:center;font-weight: bolder'
           )
        ),
        helpText(style = 'color:#000000;font-weight: normal;',
                 'For more details about the header, please', 
                 'read reference manual of the R package dplR.', 
                 'The output file is Tucson format.'),
        hr(),
        #HTML("<p style = 'color:#000000;'><b>RWL</b></p>"),
        downloadButton(
          'RingWidth.rwl', 'Download RWL',
          class = "btn btn-primary btn-md",
          style = 'color:#FFFFFF;text-align:center;font-weight:bolder;'
        )
      )
    ),
    conditionalPanel(  
      condition = 'input.tuheader',
      box(
        title = 'Header',width = 3, 
        status = 'primary', solidHeader = T, collapsible = T,
        textInput('tuhdr1', 'Site ID', ''),
        textInput('tuhdr2', 'Site Name', ''),
        textInput('tuhdr3', 'Species Code', ''),
        textInput('tuhdr4', 'State or Country', ''),
        textInput('tuhdr5', 'Species', ''),
        textInput('tuhdr6', 'Elevation', '')
      )
    ),   
    conditionalPanel(  
      condition = 'input.tuheader',
      box(
        title = 'Header',width = 3, 
        status = 'primary', solidHeader = T, collapsible = T,
        textInput('tuhdr7', 'Latitude', ''),
        textInput('tuhdr8', 'Longitude', ''),
        textInput('tuhdr9', 'First Year', ''),
        textInput('tuhdr10', 'Last Year', ''),
        textInput('tuhdr11', 'Lead Investigator', ''),
        textInput('tuhdr12', 'Completion Date', '')
      )
    )   
  )
)
shiny.body <- dashboardBody(
  tabItems(
    tabItem(tabName = 'input_pre', page1),
    tabItem(tabName = 'mea_arg', page2.1, page2.2)
  )
)
ui <- dashboardPage(
  shiny.title,
  shiny.sider,
  shiny.body
)
return(ui)
}
